contracts/fungible-tokens/andromeda-cw20/src/contract.rs

```diff
@@ -86,12 +101,108 @@ pub fn execute(ctx: ExecuteContext, msg: ExecuteMsg) -> Result<Response, Contrac
     }
 }
 
+fn lock_and_mint_factory(
+    ctx: ExecuteContext,
+    amount: Uint128,
+    factory_contract: AndrAddr,
+) -> Result<Response, ContractError> {
+    let ExecuteContext {
+        deps,
+        info,
+        env,
+        contract,
+        ..
+    } = ctx;
+
+    ensure!(
+        contract.is_contract_owner(deps.storage, info.sender.as_str())?,
+        ContractError::Unauthorized {}
+    );
+    ensure_sufficient_available_balance(deps.storage, &info.sender, amount)?;
+
+    let locked_info = FactoryInfo {
+        factory_contract: factory_contract.clone(),
+        amount: amount,
+        user: info.sender.clone(),
+    };
+
+    BALANCES.update(deps.storage, &info.sender, |balance| -> StdResult<_> {
+        Ok(balance.unwrap_or_default().checked_sub(amount)?)
+    })?;
+
+    LOCKED_TOKENS.save(deps.storage, &locked_info)?;
+
+    let factory_address = factory_contract.get_raw_address(&deps.as_ref())?;
+
+    let token_info: TokenInfoResponse =
+        deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: env.contract.address.to_string(),
+            msg: to_json_binary(&Cw20QueryMsg::TokenInfo {})?,
+        }))?;
+
+    let msg = SocketExecuteMsg::CreateDenom {
+        subdenom: token_info.name.clone(),
+        amount,
+    };
+
+    let response = wasm_execute(factory_address, &msg, vec![])?;
```

Comment:
The CW20 contract will be sending the message to the osmosis socket contract, which will be on another chain. This call implies that the contract is on the same chain.
We should handle this by sending an AMP packet to the kernel, rename `factory_address` to `osmosis_socket` address, and use `Recipient` instead of `AndrAddr` for the socket address. 
A good example of this is the `execute_send` function in our splitter contract.